symengine/printers/unicode.cpp

```diff
@@ -0,0 +1,760 @@
+    if (prec.getPrecedence(x) < precedenceEnum) {
+        auto box = apply(x);
+        box.enclose_parens();
+        return box;
+    } else {
+        return apply(x);
+    }
+}
+
+StringBox UnicodePrinter::parenthesizeLE(const RCP<const Basic> &x,
+                                         PrecedenceEnum precedenceEnum)
+{
+    Precedence prec;
+    if (prec.getPrecedence(x) <= precedenceEnum) {
+        auto box = apply(x);
+        box.enclose_parens();
+        return box;
+    } else {
+        return apply(x);
+    }
+}
+
+StringBox UnicodePrinter::apply(const RCP<const Basic> &b)
+{
+    b->accept(*this);
+    return box_;
+}
+
+StringBox UnicodePrinter::apply(const Basic &b)
+{
+    b.accept(*this);
+    return box_;
+}
+
+std::vector<std::string> init_unicode_printer_names()
+{
+    std::vector<std::string> names;
+    names.assign(TypeID_Count, "");
+    names[SYMENGINE_SIN] = "sin";
+    names[SYMENGINE_COS] = "cos";
+    names[SYMENGINE_TAN] = "tan";
+    names[SYMENGINE_COT] = "cot";
+    names[SYMENGINE_CSC] = "csc";
+    names[SYMENGINE_SEC] = "sec";
+    names[SYMENGINE_ASIN] = "asin";
+    names[SYMENGINE_ACOS] = "acos";
+    names[SYMENGINE_ASEC] = "asec";
+    names[SYMENGINE_ACSC] = "acsc";
+    names[SYMENGINE_ATAN] = "atan";
+    names[SYMENGINE_ACOT] = "acot";
+    names[SYMENGINE_ATAN2] = "atan2";
+    names[SYMENGINE_SINH] = "sinh";
+    names[SYMENGINE_CSCH] = "csch";
+    names[SYMENGINE_COSH] = "cosh";
+    names[SYMENGINE_SECH] = "sech";
+    names[SYMENGINE_TANH] = "tanh";
+    names[SYMENGINE_COTH] = "coth";
+    names[SYMENGINE_ASINH] = "asinh";
+    names[SYMENGINE_ACSCH] = "acsch";
+    names[SYMENGINE_ACOSH] = "acosh";
+    names[SYMENGINE_ATANH] = "atanh";
+    names[SYMENGINE_ACOTH] = "acoth";
+    names[SYMENGINE_ASECH] = "asech";
+    names[SYMENGINE_LOG] = "log";
+    names[SYMENGINE_LAMBERTW] = "W";
+    names[SYMENGINE_ZETA] = "\U0001D701";
+    names[SYMENGINE_DIRICHLET_ETA] = "\U0001D702";
+    names[SYMENGINE_KRONECKERDELTA] = "kroneckerdelta";
+    names[SYMENGINE_LEVICIVITA] = "levicivita";
+    names[SYMENGINE_TRUNCATE] = "truncate";
+    names[SYMENGINE_ERF] = "erf";
+    names[SYMENGINE_ERFC] = "erfc";
+    names[SYMENGINE_LOWERGAMMA] = "\U0001D6FE";
+    names[SYMENGINE_UPPERGAMMA] = "\u0393";
+    names[SYMENGINE_BETA] = "B";
+    names[SYMENGINE_LOGGAMMA] = "log \u0393";
+    names[SYMENGINE_LOG] = "log";
+    names[SYMENGINE_POLYGAMMA] = "polygamma";
+    names[SYMENGINE_GAMMA] = "\u0393";
+    names[SYMENGINE_MAX] = "max";
+    names[SYMENGINE_MIN] = "min";
+    names[SYMENGINE_SIGN] = "sgn";
+    names[SYMENGINE_CONJUGATE] = "conjugate";
+    names[SYMENGINE_PRIMEPI] = "\U0001D70B";
+    names[SYMENGINE_UNEVALUATED_EXPR] = "";
+    return names;
+}
+
+const std::vector<std::string> UnicodePrinter::names_
+    = init_unicode_printer_names();
+
+std::vector<unsigned> init_unicode_printer_lengths()
+{
+    std::vector<unsigned> lengths;
+    lengths.assign(TypeID_Count, 0);
```

Comment:
is `0` a good default value here? (e.g. if a new function is added at some point & this is not updated, maybe a larger value would be more likely to look reasonable?)

Reply:
Nice catch! Given the suggestion to default to the str printer names I'll initialize the lengths with the `string.length`. Since these strings will never be multi-byte unicode it'll work as a default.